### PR TITLE
DPMMA-902 Campaign priority

### PR DESCRIPTION
**Pull request by patrykgruszka**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ N ]
| New feature/enhancement? (use the a.x branch)      | [ Y ]
| Deprecations?                          | [ N ]
| BC breaks? (use the c.x branch)        | [ N ]
| Automated tests included?              | [ Y ] 
| Related user documentation PR URL      | https://github.com/mautic/user-documentation/pull/200
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:
This PR adds possibility to set priority of campaign (normal / high). Campaigns with higher priority will be processed before campaigns with normal priority when running the cron job.
image

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Add two campaigns with default (normal) priority
3. Run `php bin/console m:c:t` and check the order  - the oldest campaign should be executed first
4. Change the priority of the second campaign to high
5. Again run `php bin/console m:c:t` - now the second campaign should be executed first

### CODE REVIEW

**Review by annamunk — APPROVED** [review]
Nice feature! The code looks good, automated test passed, and campaigns ran in the correct order, with campaign 2 being a high priority.
campaigns_trigger

**Review by kuzmany — DISMISSED** [review]
I like this PR and works as expected. 
But I would like see icon of high priority campaigns in the list (at least) to recognize campaigns with higher priority

image

**Comment by patrykgruszka** [comment]
@kuzmany good point, I've added the icon to the Campaign list.
image

**Review by PatrickJenkner — APPROVED** [review]
Works as described for a user 👍

File: app/bundles/CampaignBundle/Tests/Command/TriggerCampaignCommandTest.php [hunk 1]
```diff
@@ -609,6 +609,30 @@ public function testSegmentCacheCount(): void
...
+
+        $campaignStartLines = [];
+        foreach ($lines as $line) {
+            if (str_starts_with($line, 'Triggering events for campaign')) {
+                $campaignStartLines[] = $line;
+            }
+        }
+
+        // check if the first processed campaign is the campaign with high priority
+        $this->assertEquals("Triggering events for campaign {$campaign->getId()}", $campaignStartLines[0]);
```
Comment (escopecz) [inline app/bundles/CampaignBundle/Tests/Command/TriggerCampaignCommandTest.php:633]: Isn't it the only campaign in the database at this moment? This test would pass even if this new feature wasn't developed, right?

**Review by escopecz — COMMENTED** [review]
I'm not actually a big fan of this approach. I understand the need for it, but adding backend processing configuration to the hands of the users is never a good idea.

This is a workaround for the real problem which is that the campaign processing is slow, right?

I would actually suggest to instead invest time in going away from cron jobs to workers. M5 will already send emails and process email hits with SymfonyMessenger. We should move all the background work to the queue which will allow us to (auto)scale the workers so users won't have to wait for anything. It will allow us to process changed contacts in (close to) real time instead of this slow and inefficient approach of rebuilding everything every few minutes. What do you think?

**Comment by kuzmany** [comment]
@escopecz I agree with you on this. If speed up processing of certain campaign it's only reason for this, then move to messenger would be a good idea.

Anyway, we are working at the moment to move `mautic:reports:scheduler` and  `mautic:custom-field:create-column`  to the messenger queue. Reports mainly for send now and fields for creation as immediately as possible on the background.

**Comment by patrykgruszka** [comment]
@escopecz, @kuzmany thank you for your opinion.
Yes, this PR is indeed a workaround for slow processing, especially for recent campaigns. I understand your concern about putting processing configuration in the hands of users. Do you have any examples or docs for worker configuration instead of cron? I would like to look into this, sounds like a good idea.

I was also wondering if we could reverse the order in which campaigns are processed, so that the most recent campaigns are processed first. With this change we would not be giving any configuration to the users and the most recent campaigns would be processed faster. What do you think about making this change instead of the priority?

File: app/bundles/CampaignBundle/Tests/Command/TriggerCampaignCommandTest.php (hunk 1 above)
Comment (patrykgruszka) [reply app/bundles/CampaignBundle/Tests/Command/TriggerCampaignCommandTest.php:633]: The first campaign is loaded suring setup (AbstractCampaignCommand::setUp), this is the second campaign.

https://github.com/mautic/mautic/blob/c0ac713ebafc38446c52325c3d0cd704489a63c1/app/bundles/CampaignBundle/Tests/Command/AbstractCampaignCommand.php#L60

**Comment by escopecz** [comment]
@patrykgruszka great idea with the order change. I see no reason why not to make this change.

We were trying to be as close to https://symfony.com/doc/5.4/messenger.html as we could. There is UI configuration on top of that.

Read through the code changes in
https://github.com/mautic/mautic/pull/11598
https://github.com/mautic/mautic/pull/11613 (how it is used in email sending)
https://github.com/mautic/mautic/pull/12509 (simplified + configuration)

**Comment by patrykgruszka** [comment]
Closing in favour of #12922

**Pull request closed without merging**